### DISCUSSION TITLE
Unable to pull ArgoCD metrics onto Grafana dashboard using Prometheus as datasource.

### DISCUSSION THREAD

**Discussion by pcnadella** [argoproj/argo-cd#8968] [Q&A]
Hello,

I am having trouble getting some useful ArgoCD metrics data onto Grafana dashboard using Prometheus as datasource after I import the ArgoCD dashboard as described [here](https://grafana.com/grafana/dashboards/14584).

<img width="1731" alt="image" src="https://user-images.githubusercontent.com/97557034/161275254-714c21a8-ceaf-40e9-9030-912d192b7ac1.png">

Do I need to enable any Prometheus related labels in ArgoCD[ values.yaml ](https://github.com/argoproj/argo-helm/blob/master/charts/argo-cd/values.yaml) to see some data coming into the dashboards?

Your help is much appreciated.

-TIA

**Comment by hansungg** [comment]
hello.
I am also having the same problem.

Have you ever solved it?
If you have solved it, please share how.

thank you.

**Comment by rbankole** [comment]
think this may be related to the servicemonitor labels if i remember correctly.

**Comment by pcnadella** [comment]
Correct, we have to deploy servicemonitors in the same namespace where the ArgoCD resides.

**Comment by reefland** [comment]
When you look at Prometheus > Status > Targets  are your serviceMonitors listed and show "UP" ? Something like this:

![prometheus_argocd_servicemonitors](https://user-images.githubusercontent.com/16230914/174898210-77d9368c-9e60-4351-80b9-1e6434b210be.png)

Prometheus ServiceMonitors are discovered by a label.  Different installation methods use different labels.  I use Kube Prometheus Stack and my serviceMonitors require the label `release=kube-prometheus-stack` to be seen.  You'll need to figure out which label your installation requires.

**Comment by ivanov-slk** [comment]
(This is old, but the below might help others like me in the future.)

I stumbled upon this thing right now and after some digging I found that the default configuration [here](https://argo-cd.readthedocs.io/en/stable/operator-manual/metrics/) needed some tweaking. The first thing was configuring the labels, as described above. 

The second was to update the `selector.matchLabels` field - I found the correct values by enabling the default service monitors in the [Helm values](https://github.com/argoproj/argo-helm/tree/e91bc78a4a337ab014adb19611dfb8f2b5ce707f/charts/argo-cd). I used the following configuration and replicated it for each of the mentioned components.
```
controller:
  metrics:
    enabled: true
    serviceMonitor:
      enabled: true
```

If you want to stick with the default ones (which seems the better choice), you would probably need to update Prometheus's configuration, so that it is enabled to look for service monitors at the correct places. In my case (using `kube-prometheus-stack`) I edited the [values.yaml](https://github.com/prometheus-community/helm-charts/blob/main/charts/kube-prometheus-stack/values.yaml#L2819). More information - in this [StackOverflow answer](https://stackoverflow.com/a/65648944/10785101).

Also, a very good resource for debugging Prometheus Service Monitors is Prometheus's [own documentation](https://github.com/prometheus-operator/prometheus-operator/blob/main/Documentation/troubleshooting.md).

If you want an example, you can check the results of my playing around [here - Prometheus](https://github.com/ivanov-slk/virtualization-set-up/tree/master/prometheus-stack) and [here - ArgoCD](https://github.com/ivanov-slk/virtualization-set-up/tree/master/argocd). Do keep in mind that I'm doing a fair bit of experimenting there, so don't expect production grade code ;)
